File: rust/crates/corlinman-mcp/src/server/dispatch.rs
//! Real `FrameHandler` implementation that drives the per-session
//! state machine and dispatches by method-prefix to a set of registered
//! [`CapabilityAdapter`]s.
//!
//! Iter 4 shipped a stub handler that always returned
//! `MethodNotFound`. Iter 9 lands the actual dispatcher so the gateway
//! integration can mount a working `/mcp` endpoint.
//!
//! ## Routing
//!
//! Method dispatch happens in three stages:
//!
//! 1. **Lifecycle gate.** [`SessionState::check_request_allowed`] /
//!    [`SessionState::check_notification_allowed`] (iter 3) refuses
//!    non-`initialize` requests while the session is still in
//!    `Connected` / `Initializing`.
//! 2. **Built-in methods.** `initialize` and
//!    `notifications/initialized` are handled here, not by an adapter:
//!    they mutate `SessionState` and emit the canonical
//!    [`InitializeResult`] reply.
//! 3. **Capability adapters.** Anything else is routed by the prefix
//!    before the first `/`. `tools/list` → the adapter whose
//!    `capability_name() == "tools"`. Adapters are stored in a
//!    `BTreeMap<&'static str, Arc<dyn CapabilityAdapter>>` so lookup
//!    is O(log n) and order is stable for snapshot tests.

use std::collections::BTreeMap;
use std::sync::Arc;

use async_trait::async_trait;
use serde_json::Value as JsonValue;
use tokio::sync::Mutex;
use tracing::{debug, warn};

use crate::adapters::{CapabilityAdapter, SessionContext};
use crate::error::McpError;
use crate::schema::{
    InitializeParams, JsonRpcRequest, JsonRpcResponse, ServerCapabilities,
};
use crate::server::session::{
    initialize_reply, SessionState, INITIALIZED_NOTIFICATION, INITIALIZE_METHOD,
};
use crate::server::transport::FrameHandler;

/// Server identity surfaced in `initialize` replies.
#[derive(Debug, Clone)]
pub struct ServerInfo {
    pub name: String,
    pub version: String,
}

impl Default for ServerInfo {
    fn default() -> Self {
        Self {
            name: "corlinman".into(),
            version: env!("CARGO_PKG_VERSION").into(),
        }
    }
}

/// Real `FrameHandler` — built from a set of capability adapters. The
/// transport's per-connection `SessionState` + `SessionContext` are
/// supplied at call time; this handler owns no per-connection state.
pub struct AdapterDispatcher {
    adapters: BTreeMap<&'static str, Arc<dyn CapabilityAdapter>>,
    server_info: ServerInfo,
    /// Server-side advertised capabilities for the `initialize` reply.
    /// Built from the adapter set at construction so adding a tools
    /// adapter automatically advertises `tools: {}`.
    capabilities: ServerCapabilities,
}

impl AdapterDispatcher {
    /// Build with no adapters; subsequent `register` calls add them.
    /// Used by tests; production callers should prefer
    /// [`AdapterDispatcher::from_adapters`].
    pub fn new(server_info: ServerInfo) -> Self {
        Self {
            adapters: BTreeMap::new(),
            server_info,
            capabilities: ServerCapabilities::default(),
        }
    }

    /// Build from a set of adapters; advertised capabilities are
    /// derived from the registered names.
    pub fn from_adapters(
        server_info: ServerInfo,
        adapters: Vec<Arc<dyn CapabilityAdapter>>,
    ) -> Self {
        let mut d = Self::new(server_info);
        for a in adapters {
            d.register(a);
        }
        d
    }

    /// Register one capability adapter. Last-write-wins on duplicates;
    /// the dispatcher logs a warning so a typo doesn't silently
    /// shadow.
    pub fn register(&mut self, adapter: Arc<dyn CapabilityAdapter>) {
        let cap = adapter.capability_name();
        if self.adapters.contains_key(cap) {
            warn!(capability = cap, "mcp dispatcher: duplicate adapter; replacing");
        }
        self.adapters.insert(cap, adapter);
        // Refresh advertised capabilities. We use the spec's "object,
        // even if empty" shape — present means supported.
        match cap {
            "tools" => {
                self.capabilities.tools = Some(crate::schema::ToolsCapability::default());
            }
            "resources" => {
                self.capabilities.resources = Some(
                    crate::schema::ResourcesCapability {
                        // C1 advertises subscribe=false (per design Open Q §3).
                        subscribe: Some(false),
                        list_changed: None,
                    },
                );
            }
            "prompts" => {
                self.capabilities.prompts = Some(crate::schema::PromptsCapability::default());
            }
            other => {
                warn!(capability = %other, "mcp dispatcher: unknown adapter capability");
            }
        }
    }

    fn capability_for(method: &str) -> Option<&'static str> {
        let prefix = method.split('/').next()?;
        match prefix {
            "tools" => Some("tools"),
            "resources" => Some("resources"),
            "prompts" => Some("prompts"),
            _ => None,
        }
    }
}

#[async_trait]
impl FrameHandler for AdapterDispatcher {
    async fn handle(
        &self,
        req: JsonRpcRequest,
        session: &Mutex<SessionState>,
        ctx: &SessionContext,
    ) -> Result<Option<JsonRpcResponse>, McpError> {
        // Notifications are gated separately — they never produce a
        // reply on the wire, even on error.
        if req.is_notification() {
            // Lifecycle gate.
            {
                let s = session.lock().await;
                if let Err(err) = s.check_notification_allowed(&req.method) {
                    debug!(method = %req.method, %err, "mcp: notification rejected by lifecycle gate");
                    return Ok(None);
                }
            }
            if req.method == INITIALIZED_NOTIFICATION {
                let mut s = session.lock().await;
                let _ = s.observe_initialized_notification();
            }
            // Other notifications are accepted but no-op in C1.
            return Ok(None);
        }

        let id = req.id.clone().unwrap_or(JsonValue::Null);

        // Lifecycle gate (requests).
        {
            let s = session.lock().await;
            if let Err(err) = s.check_request_allowed(&req.method) {
                return Err(err);
            }
        }

        // Built-in `initialize` reply.
        if req.method == INITIALIZE_METHOD {
            let parsed: InitializeParams =
                serde_json::from_value(req.params.clone()).map_err(|e| {
                    McpError::invalid_params(format!("initialize: bad params: {e}"))
                })?;
            {
                let mut s = session.lock().await;
                s.observe_initialize(&parsed)?;
            }
            let reply = initialize_reply(
                self.capabilities.clone(),
                self.server_info.name.clone(),
                self.server_info.version.clone(),
            );
            let value = serde_json::to_value(reply).map_err(|e| {
                McpError::Internal(format!("initialize: serialize reply: {e}"))
            })?;
            return Ok(Some(JsonRpcResponse::ok(id, value)));
        }

        // Capability dispatch.
        let cap = match Self::capability_for(&req.method) {
            Some(c) => c,
            None => return Err(McpError::MethodNotFound(req.method)),
        };
        let adapter = match self.adapters.get(cap) {
            Some(a) => a.clone(),
            None => return Err(McpError::MethodNotFound(req.method)),
        };
        let result_value = adapter.handle(&req.method, req.params, ctx).await?;
        Ok(Some(JsonRpcResponse::ok(id, result_value)))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adapters::CapabilityAdapter;
    use crate::schema::JsonRpcResponse;
    use serde_json::json;

    /// Dummy adapter used to exercise dispatch routing without dragging
    /// in plugin / memory / skill state.
    struct DummyAdapter {
        name: &'static str,
        last_method: std::sync::Mutex<Option<String>>,
    }

    #[async_trait]
    impl CapabilityAdapter for DummyAdapter {
        fn capability_name(&self) -> &'static str {
            self.name
        }
        async fn handle(
            &self,
            method: &str,
            _params: JsonValue,
            _ctx: &SessionContext,
        ) -> Result<JsonValue, McpError> {
            *self.last_method.lock().unwrap() = Some(method.to_string());
            Ok(json!({"adapter": self.name, "method": method}))
        }
    }

    fn dummy(name: &'static str) -> Arc<DummyAdapter> {
        Arc::new(DummyAdapter {
            name,
            last_method: std::sync::Mutex::new(None),
        })
    }

    #[tokio::test]
    async fn initialize_advertises_registered_capabilities_only() {
        let d = AdapterDispatcher::from_adapters(
            ServerInfo::default(),
            vec![dummy("tools") as Arc<dyn CapabilityAdapter>],
        );
        let session = Mutex::new(SessionState::new());
        let ctx = SessionContext::permissive();

        let req = JsonRpcRequest {
            jsonrpc: "2.0".into(),
            id: Some(json!("init-1")),
            method: "initialize".into(),
            params: json!({
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "test", "version": "0.1"}
            }),
        };
        let reply = d.handle(req, &session, &ctx).await.unwrap().expect("reply");
        match reply {
            JsonRpcResponse::Result { result, .. } => {
                assert_eq!(result["capabilities"]["tools"], json!({}));
                assert!(result["capabilities"]["resources"].is_null());
                assert!(result["capabilities"]["prompts"].is_null());
                assert_eq!(result["protocolVersion"], "2024-11-05");
                assert_eq!(result["serverInfo"]["name"], "corlinman");
            }
            other => panic!("expected Result, got {other:?}"),
        }
    }

    #[tokio::test]
    async fn pre_initialize_request_returns_session_not_initialized() {
        let d = AdapterDispatcher::from_adapters(
            ServerInfo::default(),
            vec![dummy("tools") as Arc<dyn CapabilityAdapter>],
        );
        let session = Mutex::new(SessionState::new());
        let ctx = SessionContext::permissive();

        let req = JsonRpcRequest {
            jsonrpc: "2.0".into(),
            id: Some(json!(1)),
            method: "tools/list".into(),
            params: JsonValue::Null,
        };
        let err = d.handle(req, &session, &ctx).await.expect_err("must err");
        assert_eq!(err.jsonrpc_code(), -32002); // SESSION_NOT_INITIALIZED
    }

    #[tokio::test]
    async fn post_handshake_dispatch_routes_to_capability_adapter() {
        let tools = dummy("tools");
        let d = AdapterDispatcher::from_adapters(
            ServerInfo::default(),
            vec![tools.clone() as Arc<dyn CapabilityAdapter>],
        );
        let session = Mutex::new(SessionState::new());
        let ctx = SessionContext::permissive();

        // Walk the handshake first.
        let init_req = JsonRpcRequest {
            jsonrpc: "2.0".into(),
            id: Some(json!("i")),
            method: "initialize".into(),
            params: json!({
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "t", "version": "0"}
            }),
        };
        let _ = d.handle(init_req, &session, &ctx).await.unwrap().unwrap();
        let initd = JsonRpcRequest {
            jsonrpc: "2.0".into(),
            id: None,
            method: "notifications/initialized".into(),
            params: JsonValue::Null,
        };
        let _ = d.handle(initd, &session, &ctx).await.unwrap();

        // Now dispatch.
        let req = JsonRpcRequest {
            jsonrpc: "2.0".into(),
            id: Some(json!(7)),
            method: "tools/list".into(),
            params: JsonValue::Null,
        };
        let reply = d.handle(req, &session, &ctx).await.unwrap().expect("reply");
        match reply {
            JsonRpcResponse::Result { id, result, .. } => {
                assert_eq!(id, json!(7));
                assert_eq!(result["adapter"], "tools");
                assert_eq!(result["method"], "tools/list");
            }
            other => panic!("expected Result, got {other:?}"),
        }
        assert_eq!(
            tools.last_method.lock().unwrap().as_deref(),
            Some("tools/list")
        );
    }

    #[tokio::test]
    async fn unknown_capability_returns_method_not_found() {
        let d = AdapterDispatcher::from_adapters(
            ServerInfo::default(),
            vec![dummy("tools") as Arc<dyn CapabilityAdapter>],
        );
        let session = Mutex::new(SessionState::new());
        let ctx = SessionContext::permissive();

        // Walk handshake.
        let _ = d
            .handle(
                JsonRpcRequest {
                    jsonrpc: "2.0".into(),
                    id: Some(json!("i")),
                    method: "initialize".into(),
                    params: json!({
                        "protocolVersion": "2024-11-05",
                        "capabilities": {},
                        "clientInfo": {"name": "t", "version": "0"}
                    }),
                },
                &session,
                &ctx,
            )
            .await
            .unwrap();
        let _ = d
            .handle(
                JsonRpcRequest {
                    jsonrpc: "2.0".into(),
                    id: None,
                    method: "notifications/initialized".into(),
                    params: JsonValue::Null,
                },
                &session,
                &ctx,
            )
            .await
            .unwrap();

        // resources/list with no resources adapter registered.
        let err = d
            .handle(
                JsonRpcRequest {
                    jsonrpc: "2.0".into(),
                    id: Some(json!(1)),
                    method: "resources/list".into(),
                    params: JsonValue::Null,
                },
                &session,
                &ctx,
            )
            .await
            .expect_err("must err");
        assert_eq!(err.jsonrpc_code(), -32601);
    }
}

File: python/packages/corlinman-mcp-server/src/corlinman_mcp_server/dispatch.py
"""Real :class:`FrameHandler` implementation that drives the
per-session state machine and dispatches by method-prefix to a set of
registered :class:`CapabilityAdapter`\\ s.

Mirrors the Rust ``server::dispatch`` module 1:1.

Routing happens in three stages:

1. **Lifecycle gate.**
   :meth:`SessionState.check_request_allowed` /
   :meth:`SessionState.check_notification_allowed` refuses
   non-``initialize`` requests while the session is still
   ``Connected`` / ``Initializing``.
2. **Built-in methods.** ``initialize`` and
   ``notifications/initialized`` are handled here, not by an adapter:
   they mutate :class:`SessionState` and emit the canonical
   :class:`InitializeResult` reply.
3. **Capability adapters.** Anything else is routed by the prefix
   before the first ``/``. Adapters are stored in a sorted dict so
   lookup is O(1) and iteration order is stable for snapshot tests.
"""

from __future__ import annotations

import asyncio
import os
from dataclasses import dataclass
from typing import Protocol

import structlog

from .adapters import CapabilityAdapter, SessionContext
from .errors import (
    McpError,
    McpInternalError,
    McpInvalidParamsError,
    McpMethodNotFoundError,
)
from .session import (
    INITIALIZE_METHOD,
    INITIALIZED_NOTIFICATION,
    SessionState,
    initialize_reply,
)
from .types import (
    InitializeParams,
    JsonRpcErrorResponse,
    JsonRpcRequest,
    JsonRpcResponse,
    JsonRpcResultResponse,
    PromptsCapability,
    ResourcesCapability,
    ServerCapabilities,
    ToolsCapability,
)

log = structlog.get_logger(__name__)


@dataclass
class ServerInfo:
    """Server identity surfaced in ``initialize`` replies."""

    name: str = "corlinman"
    version: str = "0.1.0"

    @classmethod
    def default(cls) -> ServerInfo:
        version = os.environ.get("CORLINMAN_MCP_SERVER_VERSION", "0.1.0")
        return cls(name="corlinman", version=version)


class FrameHandler(Protocol):
    """Pluggable JSON-RPC frame handler. The transport calls
    :meth:`handle` for every inbound frame."""

    async def handle(
        self,
        req: JsonRpcRequest,
        session: SessionState,
        session_lock: asyncio.Lock,
        ctx: SessionContext,
    ) -> JsonRpcResultResponse | JsonRpcErrorResponse | None: ...


class StubMethodNotFoundHandler:
    """Frame handler that always returns ``MethodNotFound``. Useful as
    a placeholder when standing up the transport without any
    adapters."""

    async def handle(
        self,
        req: JsonRpcRequest,
        session: SessionState,  # noqa: ARG002
        session_lock: asyncio.Lock,  # noqa: ARG002
        ctx: SessionContext,  # noqa: ARG002
    ) -> JsonRpcResultResponse | JsonRpcErrorResponse | None:
        if req.is_notification():
            return None
        raise McpMethodNotFoundError(req.method)


class AdapterDispatcher:
    """Real :class:`FrameHandler` — built from a set of capability
    adapters. The transport's per-connection :class:`SessionState` +
    :class:`SessionContext` are supplied at call time; this handler
    owns no per-connection state.

    Mirrors the Rust ``AdapterDispatcher`` 1:1.
    """

    def __init__(self, server_info: ServerInfo | None = None) -> None:
        self._server_info: ServerInfo = server_info or ServerInfo.default()
        self._adapters: dict[str, CapabilityAdapter] = {}
        self._capabilities: ServerCapabilities = ServerCapabilities()

    @classmethod
    def from_adapters(
        cls,
        server_info: ServerInfo,
        adapters: list[CapabilityAdapter],
    ) -> AdapterDispatcher:
        d = cls(server_info)
        for a in adapters:
            d.register(a)
        return d

    @property
    def server_info(self) -> ServerInfo:
        return self._server_info

    @property
    def capabilities(self) -> ServerCapabilities:
        return self._capabilities

    def register(self, adapter: CapabilityAdapter) -> None:
        """Register one capability adapter. Last-write-wins on
        duplicates; we log a warning so a typo doesn't silently
        shadow."""
        cap = adapter.capability_name()
        if cap in self._adapters:
            log.warning(
                "mcp dispatcher: duplicate adapter; replacing",
                capability=cap,
            )
        self._adapters[cap] = adapter
        # Refresh advertised capabilities. We use the spec's
        # "object, even if empty" shape — present means supported.
        if cap == "tools":
            self._capabilities.tools = ToolsCapability()
        elif cap == "resources":
            self._capabilities.resources = ResourcesCapability(subscribe=False)
        elif cap == "prompts":
            self._capabilities.prompts = PromptsCapability()
        else:
            log.warning(
                "mcp dispatcher: unknown adapter capability",
                capability=cap,
            )

    @staticmethod
    def capability_for(method: str) -> str | None:
        if "/" not in method:
            return None
        prefix = method.split("/", 1)[0]
        if prefix in ("tools", "resources", "prompts"):
            return prefix
        return None

    async def handle(
        self,
        req: JsonRpcRequest,
        session: SessionState,
        session_lock: asyncio.Lock,
        ctx: SessionContext,
    ) -> JsonRpcResultResponse | JsonRpcErrorResponse | None:
        # Notifications never produce a reply on the wire, even on
        # error.
        if req.is_notification():
            async with session_lock:
                try:
                    session.check_notification_allowed(req.method)
                except McpError as err:
                    log.debug(
                        "mcp: notification rejected by lifecycle gate",
                        method=req.method,
                        err=str(err),
                    )
                    return None
                if req.method == INITIALIZED_NOTIFICATION:
                    try:
                        session.observe_initialized_notification()
                    except McpError:
                        pass
            return None

        request_id = req.id

        async with session_lock:
            session.check_request_allowed(req.method)

        # Built-in `initialize` reply.
        if req.method == INITIALIZE_METHOD:
            try:
                parsed = InitializeParams.model_validate(req.params or {})
            except Exception as e:
                raise McpInvalidParamsError(f"initialize: bad params: {e}") from e
            async with session_lock:
                session.observe_initialize(parsed)
            reply = initialize_reply(
                self._capabilities.model_copy(deep=True),
                self._server_info.name,
                self._server_info.version,
            )
            try:
                value = reply.model_dump()
            except Exception as e:
                raise McpInternalError(f"initialize: serialize reply: {e}") from e
            return JsonRpcResponse.ok(request_id, value)

        # Capability dispatch.
        cap = self.capability_for(req.method)
        if cap is None:
            raise McpMethodNotFoundError(req.method)
        adapter = self._adapters.get(cap)
        if adapter is None:
            raise McpMethodNotFoundError(req.method)
        result_value = await adapter.handle(req.method, req.params, ctx)
        return JsonRpcResponse.ok(request_id, result_value)


__all__ = [
    "AdapterDispatcher",
    "FrameHandler",
    "ServerInfo",
    "StubMethodNotFoundHandler",
]
